Data/ByteString/Internal/Pure.hs

```diff
@@ -0,0 +1,366 @@
+putDigit !addr !off !i = pokeByteOff addr off (getDigit i)
+
+-- | Reverse bytes in the given memory range (inclusive)
+reverseBytesInplace :: Ptr Word8 -> Ptr Word8 -> IO ()
+reverseBytesInplace !p1 !p2
+  | p1 < p2 = do
+    c1 <- peekByteOff p1 0
+    c2 <- peekByteOff p2 0
+    pokeByteOff p1 0 (c2 :: Word8)
+    pokeByteOff p2 0 (c1 :: Word8)
+    reverseBytesInplace (plusPtr p1 1) (plusPtr p2 (-1))
+  | otherwise = pure ()
+
+-- | Decode signed number as decimal
+decodeSignedDec :: (Eq a, Num a, Integral a) => a -> Ptr Word8 -> IO (Ptr Word8)
+decodeSignedDec !x !buf
+  | x >= 0    = decodeUnsignedDec x buf
+  | otherwise = do
+    -- we cannot negate directly as  0 - (minBound :: Int) = minBound
+    -- So we write the sign and the first digit.
+    pokeByteOff buf 0 '-'
+    case quotRem x (-10) of
+      (q,r) -> do
+        putDigit buf 1 (fromIntegral (abs r))
+        case q of
+          0 -> pure (plusPtr buf 2)
+          _ -> decodeUnsignedDec' q (plusPtr buf 1) (plusPtr buf 2)
+
+
+-- | Decode positive number as decimal
+decodeUnsignedDec :: (Eq a, Num a, Integral a) => a -> Ptr Word8 -> IO (Ptr Word8)
+decodeUnsignedDec !v !next_ptr = decodeUnsignedDec' v next_ptr next_ptr
+
+-- | Decode positive number as little-endian decimal, then reverse it.
+--
+-- Take two pointers (orig_ptr, next_ptr) to support already decoded digits
+-- (e.g. used by decodeSignedDec to avoid overflows)
+--
+decodeUnsignedDec' :: (Eq a, Num a, Integral a) => a -> Ptr Word8 -> Ptr Word8 -> IO (Ptr Word8)
+decodeUnsignedDec' !v !orig_ptr !next_ptr = case divMod v 10 of
+  (q,r) -> do
+    putDigit next_ptr 0 (fromIntegral r)
+    case q of
+      0 -> do
+        -- reverse written digits
+        reverseBytesInplace orig_ptr next_ptr
+        -- return pointer after our digits
+        pure (plusPtr next_ptr 1)
+      _ -> decodeUnsignedDec' q orig_ptr (plusPtr next_ptr 1)
+
+decodeUnsignedDecPadded :: (Eq a, Num a, Integral a) => Int -> a -> Ptr Word8 -> IO ()
+decodeUnsignedDecPadded !max_width !v !buf = do
+  let !(q,r) = divMod v 10
+  putDigit buf (max_width - 1) (fromIntegral r)
+  case q of
+    0 -> do
+        -- pad beginning
+        let pad 0 = return ()
+            pad n = putDigit buf (n - 1) 0 >> pad (n - 1)
+        pad (max_width - 1)
+    _ -> decodeUnsignedDecPadded (max_width - 1) q buf
+    
+    
+
+-- | Decode positive number as hexadecimal
+decodeUnsignedHex :: (Eq a, Num a, Integral a, Bits a) => a -> Ptr Word8 -> IO (Ptr Word8)
+decodeUnsignedHex !v !next_ptr = decodeUnsignedHex' v next_ptr next_ptr
+
+-- | Decode positive number as little-endian hexdecimal, then reverse it.
+--
+-- Take two pointers (orig_ptr, next_ptr) to support already decoded digits
+decodeUnsignedHex' :: (Eq a, Num a, Integral a, Bits a) => a -> Ptr Word8 -> Ptr Word8 -> IO (Ptr Word8)
+decodeUnsignedHex' !v !orig_ptr !next_ptr = do
+  putDigit next_ptr 0 (fromIntegral (v .&. 0x0F))
+  let !q = v `shiftR` 4
+  case q of
```

Comment:
inline q

Reply:
I've refactored and added a comment instead